staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -2787,6 +2787,14 @@ type PodSpec struct {
 	// +k8s:conversion-gen=false
 	// +optional
 	HostIPC bool `json:"hostIPC,omitempty" protobuf:"varint,13,opt,name=hostIPC"`
+	// Use the host's user namespace.
+	// This field is alpha-level and should be enabled on a node only if user-namespace remapping feature is enabled at runtime.
+	// When NodeUserNamespace feature is enabled, value of this field defaults to false.
+	// When NodeUserNamespace feature is not enabled, this field is reset to nil from within code.
+	// Optional: Defaults to false
```

Comment:
> When NodeUserNamespace feature is enabled, value of this field defaults to false.

if a pod runs on kubernetes 1.10 today, does it use the host's user namespace? if so, doesn't that have to be the default if this field is unset? otherwise a deployment that currently worked and made use of persistent storage would be likely to break on an upgrade once this feature was enabled as new pods were created.

from the design doc:

> The main risk with this change stems from the fact that processes in Pods will run with different “real” uids than they used to, while expecting the original uids to make operations on the Nodes or consistently access shared persistent storage.

that doesn't seem acceptable to opt pods into. was there discussion on the compatibility issues caused by defaulting this on?

Reply:
> If the field is HostUserNamespace is unset and NodeUserNamespace is enabled, it should be fair to expect that user has made a conscious decision of adding extra security layer and he/she is aware that this feature does not support remote volumes.

That's not typically how feature gates are used. The intent is to disable pre-release behavior, and are generally expected to be enabled by default once that behavior is fully implemented, tested, and considered stable. In this case, enabling the `NodeUserNamespace` could not be done automatically as part of the feature promotion, since it would change the default behavior of pods and disrupt upgrade cases.

We shouldn't generally require administrators to opt out of the feature flag (manually disable a stable, on-by-default feature flag) to keep existing behavior on a future upgrade.

In the design doc, one mitigation to the noted risk was "This can be mitigated by turning the feature on gradually, per-Pod or per Kubernetes namespace." That's what I would have expected as well, but requires pods to default to their current behavior when the `hostUserNamespace` field is unset.

Reply:
In that case, to see the feature in action, user will have to explicitly set `HostUserNamespace=false`, in addition to feature enablement. Does this sound right?

Reply:
> In that case, to see the feature in action, user will have to explicitly set HostUserNamespace=false, in addition to feature enablement. Does this sound right?

opting into the changed behavior at the pod level is what I would expect.

When we get to the point of PSP integration (mentioned in the design doc as part of graduation), PSP admission would give us a place to let admins control defaulting, and enforce pods setting `hostUserNamespace=false` (very similar to the `AllowPrivilegeEscalation` field):

https://github.com/kubernetes/kubernetes/blob/b672d2ee3cf3b2fbb46e2f441e50eaaab6e7aecd/staging/src/k8s.io/api/policy/v1beta1/types.go#L187-L194

Reply:
make sense. Thanks!